### PR TITLE
[API-704] v5 SQL: Update documentation

### PR DESCRIPTION
**Pull request by mdumandag**
Updates the SQL documentation with the changes done in the reference
manual. This change mainly reorganizes the documentation and adds some
notes about the new capabilities. For the details, we directed users
to the reference manual so that we don't have to duplicate most parts
of the reference manual which can change frequently.

### CODE REVIEW

**Comment by srknzl** [comment]
@mdumandag rebase please?

**Comment by mdumandag** [comment]
@srknzl I did that, but I expect some changes in the Hazelcast reference manual after recent changes (removal of implicit mappings, json -> json-flat, etc.), so I will wait for them. You could review it now, if you want, but note that I might change a couple of things based on the updates to the Hazelcast reference manual

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 1]
```diff
@@ -1396,18 +1559,14 @@ The following code prints names of the employees whose age is less than 30:
...
 
-Querying Map
-~~~~~~~~~~~~
-
-The following subsections describe how you can access Hazelcast ``Map`` objects
-and perform queries on them.
+The following subsections describe how you can access Hazelcast maps
+and perform queries on them in more details.
 
 **Names**
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1565]: ref manual includes more info on schemas https://docs.hazelcast.com/hazelcast/5.0-snapshot/query/sql-overview.html#schemas

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 2]
```diff
@@ -1526,99 +1712,42 @@ TIMESTAMP_WITH_TIME_ZONE datetime.datetime (with non-None tzinfo)
...
+Available connectors are:
 
-Lite Members
-~~~~~~~~~~~~
+- `Apache Kafka
+  <https://docs.hazelcast.com/hazelcast/latest/sql/kafka-connector.html>`__:
+  Read from and write to Kafka topics.
+- `File
+  <https://docs.hazelcast.com/hazelcast/latest/sql/file-connector.html>`__:
+  Read from a https://docs.hazelcast.com/hazelcast/latest/sql/connectors.htmllocal or remote file.
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1743]:
```suggestion
  Read from a https://docs.hazelcast.com/hazelcast/latest/sql/connectors.html local or remote file.
```
```suggestion
  Read from a local or remote file.
```

Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1743]: again this unordered list does not look ok. It looks like this: 
a

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 3]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+
+- `SELECT <https://docs.hazelcast.com/hazelcast/latest/sql/select.html>`__:
+  Read data from a table.
+- `SINK INTO/INSERT INTO
+  <https://docs.hazelcast.com/hazelcast/latest/sql/sink-into.html>`__:
+  Ingest data into a map and/or forward data to other systems.
+- `UPDATE <https://docs.hazelcast.com/hazelcast/latest/sql/update.html>`__:
+  Overwrite values in map entries.
+- `DELETE <https://docs.hazelcast.com/hazelcast/latest/sql/delete.html>`__:
+  Delete map entries.
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1487]: again, unordered lists under `SQL Statements` does not look ok

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 4]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+- `DROP JOB <https://docs.hazelcast.com/hazelcast/latest/sql/drop-job.html>`__:
+  Cancel a job.
+- `CREATE OR REPLACE SNAPSHOT (Enterprise only)
+  <https://docs.hazelcast.com/hazelcast/latest/sql/create-snapshot.html>`__:
+  Create a snapshot of a running job, so you can stop and restart it at a
+  later date.
+- `DROP SNAPSHOT (Enterprise only)
+  <https://docs.hazelcast.com/hazelcast/latest/sql/drop-snapshot.html>`__:
+  Cancel a running job.
+
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1521]: Can we add the information in https://docs.hazelcast.com/hazelcast/5.0-snapshot/query/sql-overview.html#querying-existing-maps to here without duplicating information?

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 5]
```diff
@@ -1482,6 +1633,46 @@ from the map:
 If both the key and value have fields with the same name, then the field of
 the value is exposed.
 
+To query JSON objects or empty maps, you should create an explicit mapping
+using `CREATE MAPPING
+<https://docs.hazelcast.com/hazelcast/latest/sql/create-mapping.html>`__
+statement.
+
+For example, this code snippet creates a mapping to a new map called
+``my_map``, which stores the JSON values ``name`` and ``salary`` and query
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1642]:
```suggestion
``employees``, which stores the JSON values ``name`` and ``salary`` and query
```

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 6]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+  Cancel a job.
+- `CREATE OR REPLACE SNAPSHOT (Enterprise only)
+  <https://docs.hazelcast.com/hazelcast/latest/sql/create-snapshot.html>`__:
+  Create a snapshot of a running job, so you can stop and restart it at a
+  later date.
+- `DROP SNAPSHOT (Enterprise only)
+  <https://docs.hazelcast.com/hazelcast/latest/sql/drop-snapshot.html>`__:
+  Cancel a running job.
+
+Querying Map
```
Comment (srknzl) [inline docs/using_python_client_with_hazelcast_imdg.rst:1522]: The information under this is left from 4.2 I guess. And links are pointing to imdg while the ones above point to hazelcast.

https://docs.hazelcast.com/hazelcast/latest/sql/drop-snapshot.html

vs 

https://docs.hazelcast.com/imdg/latest/sql/querying-imap.html#key-and-value-fields

New hazelcast docs has [a new section](https://docs.hazelcast.com/hazelcast/5.0-snapshot/query/querying-maps-sql.html) for Querying map with SQL. Should we delete this section and add the new info here? There may be some outdated info, I did not compare then in detail.

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 7]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+Supported Queries
+~~~~~~~~~~~~~~~~~
+
+With SQL you can run various types of queries to get insight from your data:
+
+- Ad-hoc queries, also known as point queries or online transactional
+  processing (OLTP) queries.
+- Batch queries, also known as online analytical processing (OLAP) queries.
+- Streaming queries, also known as continuous queries.
+- Federated queries.
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1370]: They work & display fine on my local. Make sure that you have the dependencies installed (pip install -r requirements-dev.txt) and you generate the documentation using `make html` on the docs folder

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 8]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+with a filter on it. In SQL, a stream is like a table with infinitely many rows
+which you can only access sequentially and thus never reach the end. For
+example, you will get an error if you try to aggregate the whole stream.
+
+**Federated Queries**
 
-Consider that we have a map called ``emp`` that contains values of type
+Federated queries are those that join tables from different datasets. Normally,
+querying in SQL is database or dataset-specific. However Hazelcast comes with
+built-in connectors to allow you to pull information from different sources
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1430]: Not a big deal. I find my version a bit better because we link to the connectors page when we actually describe what we support

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 9]
```diff
@@ -1349,19 +1349,182 @@ the entry processor in the ``Map`` methods. See the following example.
...
+
+.. code:: sql
+
+    SELECT
+        trades.ticker, trades.amount, companies.name
+    FROM
+        trades
+    JOIN companies
+    ON trades.ticker = companies.ticker;
+
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1473]: No, I intentionally omit them. In our documentation structure, the next section is already the SQL statements. So, I don't think we need a link here.

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 1 above)
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1565]: This section is actually not correct anymore. Let me leave this like this and update it once the reference manual is updated

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 4 above)
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1521]: No, the info here is already in the next section. I don't see a point in doing that

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 6 above)
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1522]: No need to delete it. We just need to update the content when the up-to-date version of the reference manual is written. There are still some wrong information on the reference manual and I don't want to update this prior to that.

Think of this PR as a piece of a bigger update. The final and correct version will be sent in a separate PR.

I will update the single link pointing to IMDG reference manual BTW.

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/using_python_client_with_hazelcast_imdg.rst:1375]: Still the same. I tried with chrome and firefox. I am using virtualenv of python 3.8

sphinx version 3.4.2

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/using_python_client_with_hazelcast_imdg.rst:1484]: I agree with links to statements. We may include the equivalent query though.

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/using_python_client_with_hazelcast_imdg.rst:1375]: Sorry I think I was using python2 virtualenv. Let me try again

[the remark this answers is not in the document]
Comment (srknzl) [reply docs/using_python_client_with_hazelcast_imdg.rst:1375]: See this I think there is an issue with newer versions of docutils. If you reinstall or upgrade docutils you will experience the same as well. Your docutils version is old, that's why it works on your local. To solve this we can upgrade sphinx-rtd-theme to v1 which includes a fix for docutils 0.17.1. https://github.com/readthedocs/sphinx_rtd_theme/pull/1185

**Review by srknzl — APPROVED** [review]
Approving but I hope you add a fix for my comment https://github.com/hazelcast/hazelcast-python-client/pull/461#discussion_r708283069

File: docs/using_python_client_with_hazelcast_imdg.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:1375]: OK, I have upgraded the sphinx and theme versions. Can you check against it? These versions also produced correct looking documentations for me as well

[the remark this answers is not in the document]
Comment (srknzl) [reply docs/using_python_client_with_hazelcast_imdg.rst:1375]: yes it works now
